File: practice/cpp/algorithms.cpp
// ============================================================
// Practice: Algorithms
// Summary: linear search, binary search, Fibonacci (iter + rec).
// Run:     g++ -std=c++17 practice/cpp/algorithms.cpp -o practice/cpp/algorithms && practice/cpp/algorithms
// ============================================================

#include <iostream>
#include <vector>

// ----- 1. Linear search -> index, or -1 if not found -----
int linearSearch(const std::vector<int>& arr, int target) {
    for (int i = 0; i < (int)arr.size(); i++)
        if (arr[i] == target) return i;
    return -1;
}

// ----- 2. Binary search on a SORTED vector -> index, or -1 -----
int binarySearch(const std::vector<int>& arr, int target) {
    int lo = 0, hi = (int)arr.size() - 1;
    while (lo <= hi) {
        int mid = (lo + hi) / 2;
        if (arr[mid] == target) return mid;
        else if (arr[mid] < target) lo = mid + 1;
        else hi = mid - 1;
    }
    return -1;
}

// ----- 3a. Fibonacci, iterative (N terms) -----
std::vector<long long> fibIterative(int n) {
    std::vector<long long> seq;
    long long a = 0, b = 1;
    for (int i = 0; i < n; i++) {
        seq.push_back(a);
        long long next = a + b;
        a = b;
        b = next;
    }
    return seq;
}

// ----- 3b. Fibonacci, recursive (nth term) -----
long long fibRecursive(int n) {
    if (n < 2) return n;
    return fibRecursive(n - 1) + fibRecursive(n - 2);
}

void printVec(const std::vector<long long>& v) {
    std::cout << "[";
    for (size_t i = 0; i < v.size(); ++i) {
        if (i) std::cout << ", ";
        std::cout << v[i];
    }
    std::cout << "]" << std::endl;
}

int main() {
    std::cout << linearSearch({4, 2, 7, 1, 9, 3}, 7) << std::endl;  // 2
    std::cout << linearSearch({4, 2, 7, 1, 9, 3}, 5) << std::endl;  // -1
    std::cout << binarySearch({1, 3, 5, 7, 9, 11}, 7) << std::endl; // 3
    std::cout << binarySearch({1, 3, 5, 7, 9, 11}, 4) << std::endl; // -1
    printVec(fibIterative(10));                                      // [0, 1, 1, 2, 3, 5, 8, 13, 21, 34]
    std::vector<long long> rec;
    for (int i = 0; i < 10; i++) rec.push_back(fibRecursive(i));
    printVec(rec);                                                   // same
    return 0;
}

File: practice/rust/algorithms.rs
// ============================================================
// Practice: Algorithms
// Summary: linear search, binary search, Fibonacci (iter + rec).
// Run:     rustc practice/rust/algorithms.rs -o practice/rust/algorithms && practice/rust/algorithms
// ============================================================

// ----- 1. Linear search -> index, or -1 if not found -----
fn linear_search(arr: &[i32], target: i32) -> i32 {
    for (i, &x) in arr.iter().enumerate() {
        if x == target {
            return i as i32;
        }
    }
    -1
}

// ----- 2. Binary search on a SORTED slice -> index, or -1 -----
fn binary_search(arr: &[i32], target: i32) -> i32 {
    let (mut lo, mut hi) = (0i32, arr.len() as i32 - 1);
    while lo <= hi {
        let mid = (lo + hi) / 2;
        let val = arr[mid as usize];
        if val == target {
            return mid;
        } else if val < target {
            lo = mid + 1;
        } else {
            hi = mid - 1;
        }
    }
    -1
}

// ----- 3a. Fibonacci, iterative (N terms) -----
fn fib_iterative(n: usize) -> Vec<u64> {
    let mut seq = Vec::new();
    let (mut a, mut b) = (0u64, 1u64);
    for _ in 0..n {
        seq.push(a);
        let next = a + b;
        a = b;
        b = next;
    }
    seq
}

// ----- 3b. Fibonacci, recursive (nth term) -----
fn fib_recursive(n: u64) -> u64 {
    if n < 2 {
        return n;
    }
    fib_recursive(n - 1) + fib_recursive(n - 2)
}

fn main() {
    println!("{}", linear_search(&[4, 2, 7, 1, 9, 3], 7));   // Output: 2
    println!("{}", linear_search(&[4, 2, 7, 1, 9, 3], 5));   // Output: -1
    println!("{}", binary_search(&[1, 3, 5, 7, 9, 11], 7));  // Output: 3
    println!("{}", binary_search(&[1, 3, 5, 7, 9, 11], 4));  // Output: -1
    println!("{:?}", fib_iterative(10));                      // [0, 1, 1, 2, 3, 5, 8, 13, 21, 34]
    let rec: Vec<u64> = (0..10).map(fib_recursive).collect();
    println!("{:?}", rec);                                    // same
}
